toolchain/check/import_cpp.cpp

```diff
@@ -70,13 +68,81 @@ auto ImportCppFiles(
         IntAsSelect, IntAsSelect, IntAsSelect, std::string);
     context.emitter().Emit(loc, CppInteropParseError, num_errors, num_warnings,
                            num_imports, diagnostics_str);
-  } else if (num_warnings > 0) {
+    return std::nullopt;
+  }
+  if (num_warnings > 0) {
     CARBON_DIAGNOSTIC(CppInteropParseWarning, Warning,
                       "{0} warning{0:s} in `Cpp` {1} import{1:s}:\n{2}",
                       IntAsSelect, IntAsSelect, std::string);
     context.emitter().Emit(loc, CppInteropParseWarning, num_warnings,
                            num_imports, diagnostics_str);
   }
+  return ast;
+}
+
+static auto AddNamespace(Context& context, IdentifierId cpp_package_id,
+                         SemIR::InstId first_import_decl_id) -> void {
+  // TODO: Extract and deudplicate common logic with import.cpp.
+  auto name_id = SemIR::NameId::ForIdentifier(cpp_package_id);
+  SemIR::TypeId namespace_type_id =
+      context.GetSingletonType(SemIR::NamespaceType::SingletonInstId);
+
+  auto* parent_scope = &context.name_scopes().Get(SemIR::NameScopeId::Package);
+  auto [inserted, entry_id] = parent_scope->LookupOrAdd(
+      name_id,
+      // This InstId is temporary and would be overridden if used.
+      first_import_decl_id, SemIR::AccessKind::Public);
+  CARBON_CHECK(inserted);
+
+  CARBON_CHECK(first_import_decl_id.has_value());
+  auto import_loc_id = context.insts().GetLocId(first_import_decl_id);
+
+  auto namespace_inst = SemIR::Namespace{
+      namespace_type_id, SemIR::NameScopeId::None, first_import_decl_id};
+
+  auto namespace_inst_and_loc =
+      import_loc_id.is_import_ir_inst_id()
+          ? context.MakeImportedLocAndInst(import_loc_id.import_ir_inst_id(),
+                                           namespace_inst)
+          // TODO: Check that this actually is an `AnyNamespaceId`.
+          : SemIR::LocIdAndInst(Parse::AnyNamespaceId(import_loc_id.node_id()),
+                                namespace_inst);
+
+  auto namespace_id =
+      context.AddPlaceholderInstInNoBlock(namespace_inst_and_loc);
+
+  context.import_ref_ids().push_back(namespace_id);
+  namespace_inst.name_scope_id = context.name_scopes().Add(
+      namespace_id, name_id, SemIR::NameScopeId::Package);
+  context.ReplaceInstBeforeConstantUse(namespace_id, namespace_inst);
+
+  // Note we have to get the parent scope freshly, creating the imported
```

Comment:
Maybe we should just avoid the local variable entirely? It's only used once each time. And leave this comment around explaining why we don't hold a local variable instead?

Reply:
Done.